controller/engine_controller.go

```diff
@@ -1768,30 +1781,30 @@ func (ec *EngineController) startRebuilding(e *longhorn.Engine, replica, addr st
 			// it is past the Backoff period, we'll try to mark the Replica as Failed and increase the Backoff period
 			// for the next failure.
 			if !ec.backoff.IsInBackOffSinceUpdate(e.Name, time.Now()) {
-				rep, err := ec.ds.GetReplica(replica)
-				if err != nil {
-					log.WithError(err).Errorf("Failed to get replica %v unable to mark failed rebuild", replica)
+				if err := ec.updateReplicaRebuildFailedCondition(replica, replicaRebuildErrMsg); err != nil {
+					log.WithError(err).Errorf("Failed to update rebuild status information on replica %v", replicaName)
 					return
 				}
-				rep.Spec.FailedAt = util.Now()
-				rep.Spec.DesireState = longhorn.InstanceStateStopped
-				if _, err := ec.ds.UpdateReplica(rep); err != nil {
-					log.WithError(err).Errorf("Unable to mark failed rebuild on replica %v", replica)
+
+				replica.Spec.FailedAt = util.Now()
+				replica.Spec.DesireState = longhorn.InstanceStateStopped
+				if _, err := ec.ds.UpdateReplica(replica); err != nil {
+					log.WithError(err).Errorf("Unable to mark failed rebuild on replica %v", replicaName)
```

Comment:
If somehow the replica update failed, but the previous condition has been updated, so it will cause inconsistent temporarily; however, it should be eventually consistent. 

This looks like we can' make the spec and status update at once to cause potential non-atomic translation update.

Reply:
Removing 'getting the newest replica' will cause replica update failed and get into an infinite loop. Just Put it back.

Reply:
Exactly, because our implementation is not always atomic :slightly_frowning_face: and this will require us to get the new resource update before the next update.